kvbc/src/kvbc_adapter/blocks_deleter_adapter.cpp

```diff
@@ -0,0 +1,83 @@
+#include "IntervalMappingResourceManager.hpp"
+
+using concord::performance::ISystemResourceEntity;
+
+namespace concord::kvbc::adapter {
+
+BlocksDeleterAdapter::BlocksDeleterAdapter(
+    const std::shared_ptr<concord::storage::rocksdb::NativeClient> &native_client,
+    bool link_st_chain,
+    const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>> &category_types,
+    const std::optional<aux::AdapterAuxTypes> &aux_types)
+    : kvbc_(std::make_shared<concord::kvbc::categorization::KeyValueBlockchain>(
+          native_client, link_st_chain, category_types)) {
+  if (aux_types.has_value()) {
+    m_state_transfer_ = aux_types->state_transfer_ptr_;
+    replica_resources_.reset(&(aux_types->resource_entity_));
+    kvbc_->setAggregator(aux_types->aggregator_);
+  }
+  ConcordAssertEQ(!kvbc_, false);
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// IBlocksDeleter implementation
+void BlocksDeleterAdapter::deleteGenesisBlock() {
+  const auto genesisBlock = kvbc_->getGenesisBlockId();
+  if (genesisBlock == 0) {
+    throw std::logic_error{"Cannot delete the genesis block from an empty blockchain"};
+  }
+  kvbc_->deleteBlock(genesisBlock);
+}
+BlockId BlocksDeleterAdapter::deleteBlocksUntil(BlockId until) {
+  auto mes_prun_util(replica_resources_ ? std::make_unique<ISystemResourceEntity::scopedDurMeasurment>(
+                                              *replica_resources_, ISystemResourceEntity::type::pruning_utilization)
+                                        : nullptr);
+  const auto genesisBlock = kvbc_->getGenesisBlockId();
+  if (genesisBlock == 0) {
+    throw std::logic_error{"Cannot delete a block range from an empty blockchain"};
+  } else if (until <= genesisBlock) {
+    throw std::invalid_argument{"Invalid 'until' value passed to deleteBlocksUntil()"};
+  }
+
+  // Inform State Transfer about pruning. We must do it in this thread context for persistency considerations, and in
+  // this layer to lower the chance for bugs (there are multiple callers to this function), in which pruning is not
+  // notified to ST. In that case, ST state will be corrupted.
+  if (m_state_transfer_ && m_state_transfer_->isRunning()) {
+    // We assume until > 0, see check above
+    m_state_transfer_->reportLastAgreedPrunableBlockId(until - 1);
+  }
+
+  const auto lastReachableBlock = kvbc_->getLastReachableBlockId();
+  const auto lastDeletedBlock = std::min(lastReachableBlock, until - 1);
+  const auto start = std::chrono::steady_clock::now();
+  for (auto i = genesisBlock; i <= lastDeletedBlock; ++i) {
+    auto mes_prun_avg(replica_resources_ ? std::make_unique<ISystemResourceEntity::scopedDurMeasurment>(
+                                               *replica_resources_, ISystemResourceEntity::type::pruning_avg_time_micro)
+                                         : nullptr);
+    ConcordAssert(kvbc_->deleteBlock(i));
+  }
```

Comment:
eventually just this code is implementation sepecific per storage implementation.
The rest of the code can be in a common wrapper.
Maybe it should be extracted to the replica.cpp ?

Reply:
Thanks @cloudnoize
I think you are correct. This code contains the notion of genesis block and such notions are specific to blockchain implementations. So i am thinking of moving the following into Replica.cpp:
// Inform State Transfer about pruning. We must do it in this thread context for persistency considerations, and in
  // this layer to lower the chance for bugs (there are multiple callers to this function), in which pruning is not
  // notified to ST. In that case, ST state will be corrupted.
  if (m_state_transfer_ && m_state_transfer_->isRunning()) {
    // We assume until > 0, see check above
    m_state_transfer_->reportLastAgreedPrunableBlockId(until - 1);
  }